src/main/scala/com/datasonnet/PortX.scala

```diff
@@ -31,10 +32,10 @@ object PortX {
           datetimeObj.format(DateTimeFormatter.ofPattern(outputFormat))
       },
 
-      builtin("compare", "datetime1", "format1", "datetime2", "format2") {
-        (ev, fs, datetime1: String, format1: String, datetime2: String, format2: String) =>
-          val datetimeObj1 = java.time.ZonedDateTime.parse(datetime1, DateTimeFormatter.ofPattern(format1))
-          val datetimeObj2 = java.time.ZonedDateTime.parse(datetime2, DateTimeFormatter.ofPattern(format2))
+      builtin0("compare", "datetime1", "format1", "datetime2", "format2") {
+        (vals, ev, fs) =>
+          val datetimeObj1 = java.time.ZonedDateTime.parse(vals(0).cast[Val.Str].value, DateTimeFormatter.ofPattern(vals(1).cast[Val.Str].value))
```

Comment:
Replacing the use of builtin with builtin0 and these casts gets rid of the pretty type checking in `validate`, so I'd like to find a way to keep that (here and below)

Reply:
Ah!

A couple of findings, and you probably already knew some of this... 
1. The validate method checks that the variables are of the right type before passing the values to the custom builtin, providing some sort of compile/runtime check.
2. If you use `.cast[Val.Str]` instead of `.asInstanceOf[Val.Str]` with `builtin0` or `builtinWithDefaults` you actually get a similar message. This is from validate
```
Problem executing map: sjsonnet.Error: Wrong parameter type: expected String, got number
    at line 1 column 25 of the transformation
```

vs this one from using `cast`

```
Problem executing map: sjsonnet.Error: Expected string, found number
    at line 1 column 25 of the transformation
```

3. Related to 2. If you do something like `DS.Formats.read(payload, 123)` then you get `ClassCastException` instead of the pretty exception because it's using `asInstanceOf`.

Now while the `cast` message is as useful as the one from validate it's probably better if the user does not see two different messages for what's really the same type of error. So I'll make some changes to manually call validate from wherever is appropriate.

Reply:
Oh, that's really interesting, both how close the error message is, and how it isn't the same. Since it sounds like you'll be able to get it the same, I'll wait on that before merging, but if for some reason that turns out to not be easy to do, I'm fine with merging since the error is still readable. Btw, thanks for pointing out what happens with asInstanceOf, fixing up error handling on our other defined functions is medium high on my TODOs.